File: mathewDavidov/assignments/fastSlow/lc141/lc141.java
/**
 * Given head, the head of a linked list, determine if the linked list has a
 * cycle in it. There is a cycle in a linked list if there is some node in the
 * list that can be reached again by continuously following the next pointer.
 * Internally, pos is used to denote the index of the node that tail's next
 * pointer is connected to. Note that pos is not passed as a parameter. Return
 * true if there is a cycle in the linked list. Otherwise, return false.
 * 
 * Edge cases:
 * null head -> false
 * head points to null -> false
 * 
 * Initialize fast and slow pointers that point to the head
 * Iterate while the fast pointer is not null and does not point to null (so we can reference the next's next)
 *   Move the fast pointer reference to its next next pointer
 *   Move the slow pointer reference to its next 
 *   If the references are equal (they both reference the same node), return true, as there is a cycle
 * Return false, as the fast pointer has hit a null reference, as there is no cycle
 * 
 * Runtime:
 * Time: O(n)
 * Space: O(1)
 */

class ListNode {
    public int value = 0;
    ListNode next;

    ListNode(int value) {
        this.value = value;
        next = null;
    }
};

public class lc141 {
    public static boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }

        ListNode fast = head, slow = head;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;

            if (fast == slow) {
                return true;
            }
        }

        return false;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);
        head.next.next.next.next.next = new ListNode(6);
        System.out.println("LinkedList has cycle: " + hasCycle(head));

        head.next.next.next.next.next.next = head.next.next;
        System.out.println("LinkedList has cycle: " + hasCycle(head));

        head.next.next.next.next.next.next = head.next.next.next;
        System.out.println("LinkedList has cycle: " + hasCycle(head));
    }
}

File: joshuaCarpentier/assignments/fastSlow/lc141/lc141.js
// Problem Statement #

// Given the head of a Singly LinkedList, write a function to determine if the LinkedList has a cycle in it or not.

/**
 * Input: Starting from the head we have a fast and slow pointer
 * 
 * then as we iterate through the linked list we are changing the values of each pointer to the point that they are moving at different rates 
 * the fast pointer moves two spots while the slow moves one spot ahead meaning that the fast pointer includes some recursion 
 * 
 * and as we move towards the end we set the last pointer of the linked list to another part of the linked list 
 * specifically where the fast pointer would be meaning if we have for example :
 * 
 * 3 2 0 4
 * 
 * 3 -> 2 -> 0 -> null
 * 
 * slow
 * fast
 * 3 -> 2 -> 0 -> null
 * 
 *    slow
 *         fast 
 * 3 -> 2 -> 0 -> null
 * 
 * fast moves at 2 spot so if loop it should end up a 2 so we pointer at 2
 * 
 *         slow
 *               fast 
 * 3 -> 2 -> 0 -> |
 *      /\        |
 *      ||        |
 *      ||________| 
 *     fast
 * 
 * so we can loop through the linked list until fast and slow pointers
 * match position 
 * 
 * 
 * output: and return true if the match else false
 */

class Node {
  constructor(value, next=null){
    this.value = value;
    this.next = next;
  }
}

const has_cycle = function(head) {
  let slow = head,
      fast = head;

      while (fast !== null && fast.next !== null){
        fast = fast.next.next;
        slow = slow.next;
        if (slow === fast){
          return true;
        }
      }

// dont necessarly need to have else when just returning boolean value
  return false
}

// head is global vairiable since were not declaring with keywords like
// let, const, or var which isnt recommend 
console.log(head = new Node(1)) // adding intial node 
console.log(head.next = new Node(2))
head.next.next = new Node(3)
head.next.next.next = new Node(4)
head.next.next.next.next = new Node(5)
head.next.next.next.next.next = new Node(6)
console.log(head)
console.log(`LinkedList has cycle: ${has_cycle(head)}`)

head.next.next.next.next.next.next = head.next.next
console.log(`LinkedList has cycle: ${has_cycle(head)}`)

head.next.next.next.next.next.next = head.next.next.next
console.log(`LinkedList has cycle: ${has_cycle(head)}`)


// Time Complexity #
// As we have concluded above, once the slow pointer enters the cycle, the fast pointer will meet the slow pointer in the same loop. Therefore, the time complexity of our algorithm will be O(N) where ‘N’ is the total number of nodes in the LinkedList.

// Space Complexity #
// The algorithm runs in constant space O(1).
